### PR TITLE
feat(lua): vim.ui_attach to get ui events from lua

### PR DESCRIPTION
**Pull request by famiu**
Supersedes #16396

### CODE REVIEW

**Comment by lewis6991** [comment]
I know the PR is still draft, but is there a reason this is done for lua only as opposed to the API?

**Comment by famiu** [comment]
> I know the PR is still draft, but is there a reason this is done for lua only as opposed to the API?

It takes a Lua callback

**Comment by clason** [comment]
Done except for docs; suggestions welcome.

File: src/nvim/lua/executor.c [hunk 1]
```diff
@@ -583,6 +586,89 @@ static bool nlua_init_packages(lua_State *lstate)
...
+  int nargs = lua_gettop(lstate);
+
+  if (nargs != 3) {
+    return luaL_error(lstate, "Expected 3 arguments");
+  }
+  if (!lua_isnumber(lstate, 1)) {
+    return luaL_error(lstate, "ns_id must be an integer");
+  }
+
+  uint32_t ns_id = (uint32_t)lua_tointeger(lstate, 1);
```
Comment (bfredl) [inline src/nvim/lua/executor.c:602]: can use `luaL_checkinteger`

Comment (famiu) [reply src/nvim/lua/executor.c]: done

**Comment by bfredl** [comment]
merged the changes into https://github.com/neovim/neovim/pull/16396 and began docs/tests

**Pull request closed without merging**